# 2. Python en acción: tus primeros pasos

Python no solo es un lenguaje de programación poderoso, sino también una herramienta que podemos utilizar de inmediato, sin configuraciones complicadas. En este módulo, daremos nuestros primeros pasos explorando cómo Python puede utilizarse como una calculadora interactiva, permitiéndonos realizar operaciones matemáticas de manera sencilla y directa.

Pero Python no se limita a hacer cálculos. También permite **crear o definir elementos digitales** que podrás usar para organizar la información cuando construimos nuevos programas. Esto es fundamental porque nos da la posibilidad de reutilizar valores, modificarlos, combinarlos en expresiones más complejas y ¡hacerlos interactuar entre sí!. Esta característica, que iremos estudiando en estos módulos, nos permitirá crear programas de la complejidad que se nos antoje (y bueno, es una de las razones por la cual Python es tan popular y útil), pero ten en cuenta que comenzaremos aprendiendo con los elementos más sencillos disponibles tales como números o texto. Estudiaremos cómo funcionan y a diferenciarlos, sentando las bases que nos facilitarán la creación de programas más sofisticados más adelante.


## Metodología del tutorial 

A partir de ahora, cada explicación del tutorial estará usualmente acompañada de casillas de código que deberás inspeccionar (o estudiar) y ejecutar. Tu objetivo siempre debe ser **comprender por qué se produce cada resultado**, de modo que siempre debes tratar de **interpretar los códigos presentados intentando predecir cuál será el resultado de cada ejecución**. Para lograrlo, ¡podrás modificar a tu gusto los ejemplos y verificar sus resultados nuevos! Si tu predicción del resultado de una ejecución resulta correcta, entonces ¡estás aprendiendo el lenguaje! Puedes desarrollar la cantidad de ejemplos que gustes hasta alcanzar el aprendizaje.

También te encontrarás lugares donde se te pedirá, como ejercicio, que desarrolles algún un nuevo ejemplo. **¡No te saltes estos desafíos!** Están pensados para que debas pensar, deducir o construir algo que es clave en tu aprendizaje del lenguaje. Aprender Python es similar a aprender un idioma, deporte, o instrumento musical nuevo: solo leyendo no es suficiente para que puedas hablar el idioma, jugar el deporte, tocar el instrumento o programar en Python. **Es fundamental que pongas en práctica tu conocimiento desde tu mente, tratando de explotar toda tu creatividad**. Te recomiendo que en esta primera etapa te apoyes lo menos posibles en buscadores como Google o IAs (como ChatGPT o Blackbox) que te desarrollen los ejemplos, dado que aquellos serán _simples_ o _poco creativos_, de modo que no te ayudarán a aprender en serio. ¡Desafía tu imaginación! te garantizo que eso marcará la diferencia en tu aprendizaje. 


## 2.1 Python como calculadora

Python puede usarse como una **calculadora interactiva**. Es decir, puedes crear instrucciones que le pidan al _intérprete_ realizar operaciones matemáticas de manera sencilla y directa. Podemos sumar (`+`), restar (`-`), multiplicar (`*`), dividir (`/`) ¡y muchas más!. Prueba estos ejemplos (recuerda que las celdas se ejecutan con `Shift` + `Enter`):

In [ ]:
3+5

8

In [7]:
42.82-14.57

28.25

In [6]:
14*0.6

8.4

In [5]:
1/9

0.1111111111111111

También puede combinar operaciones de forma correcta. Python respeta el uso de paréntesis y el orden estándar en que las operaciones deben evaluarse a nivel matemático. Por ejemplo: 

In [2]:
3*(6/3 + 1)

9.0

Es decir, Python es capar de evaluar correctamente todos esos ejemplos de memes que circulan por las redes sociales tratando de desafiar a las personas mezclando operaciones matemáticas elementales que deben ser evaluadas siguiendo el orden establecido por las reglas matemáticas. ¡Compruébalo! Aquí está el clásico "desafío" que hasta se ganó [una nota en un portal de noticias]([https://tn.com.ar/tecno/juegos/2024/11/29/cuanto-es-100-10-5-x-2-50-el-desafio-matematico-solo-para-genios/):

In [9]:
100+10/5*2+50

154.0

**Ejercicio**: A continuación tienes algunas celdas de código donde debes crear tres ejemplos _creativos_ que utilicen y/o combinen operaciones matemáticas simples y paréntesis que tú puedas calcular manualmente. ¡Comprueba que el resultado es correcto!

Python también **puede realizar cálculos más complejos**, por ejemplo, usando potencias (`**`), raíces y una infinidad de cálculos más avanzados (hasta matemáticas de niveles universitarios...). Lo mejor de todo es que Python evalúa las expresiones en tiempo real, mostrando el resultado de inmediato, lo que lo convierte en una herramienta ideal para explorar y probar cálculos rápidamente. 

Acá hay algunos ejemplo con potencias. ¿Podrías calcular los resultados a mano antes de ejecutar las celdas? Si no, ¡no importa! Python te dará igualmente el resultado correcto:

In [ ]:
4**2

In [ ]:
2**4

In [ ]:
2**3 + 3**2 

Ahora algo un poco más desafiante para el cálculo manual (desafío digno de meme, pero más interesante que el anterior)...

In [10]:
2*(6/3 + 1)**3

54.0

Si te gustan las matemáticas, te habrás dado cuenta que hasta ahora hemos visto ejemplos que son posibles de resolver a mano (o incluso mentalmente). No obstante, Python nos ofrece la posibilidad de calcular cosas que mentalmente serían un lío... Pensemos en el ejemplo anterior, pero ahora suponiendo que necesitamos que el exponente de la potencia sea un número monstruosamente grande (e elegido 456 por azar). ¿Qué resultado te daría el intérprete de esta operación?

In [11]:
2*(6/3 + 1)**456

7.384517904671098e+217

¡Y violà! Python te entregó la respueta con la misma rapidez y confianza que con los ejemplos simples. ¡El resultado es número que tiene 218 cifras! El cual lamentablemente no se puede "mostrar" completo por las limitaciones tecnológicas de este tipo de operaciones. En cambio, acá se utiliza la famosa _notación científica_ para representarlo, en la cual se muestran las primeras cifras del número, y se agrupan todas las otras utilizando una potencia de 10. Esto último es indicado con la letra `e` (tanto mayúscula como minúscula) en este lenguaje. Es decir, el número anterior se podría escribir como $7,384 \times 10^{217}$.

En general, se puede utilizar esta `e` para escribir números extremadamente grandes (o pequeños) utilizando esta notación científica. ¡Python no tendrá problemas en operar con ellos tal como lo hace con los otros!, aunque considerando algunas caractarísticas distintas que estudiaremos más adelante. ¿Te quedan dudas de cómo funciona esta forma de escribir números grandes? Observa este ejemplo que simplemente escribe el número _un millón_ utilizándola:

In [14]:
1e6

1000000.0

O algo más entretenido, el número aproximado de estrellas que hay en nuestra galaxia:

In [15]:
1e10

10000000000.0

Para el caso de los números muy chiquitos podemos usar la misma notación. Por ejemplo, consideremos el ancho de un cabello humano, aproximadamente 0.04 milímetros (mm). Para escribir este valor en metros, tenemos que dividirlo por 1000, la cantidad de milímetros que hay en un metro. Esto sería:

In [6]:
0.04/1000

4e-05

Este resultado, $4\text{e-}05$, es equivalente a escribir $4 \times 10^{-5}$, que es la notación científica para el número $0.00004$.

¡Pero esto no es todo! Python también incluye algunas operaciones _interesantes_, que no las solemos aplicar en nuestra matemática académica, pero si bastante en la vida cotidiana. Por ejemplo, tiene una forma de calcular el **resto de la división**, un valor muy trascendente cuando realizamos una dividisión manualmente, y que tiene una variedad de aplicaciones dentro de la programación:

In [7]:
6%4

2

## 2.2 Celdas con múltples instrucciones

Hasta ahora hemos utilizado las celdas de código de una forma sencilla: con una única instrucción en Python que nos ha permitido ejecutar operaciones simples y obtener resultados de inmediato. Sin embargo, los _notebooks_ nos permiten incluir **múltiples instrucciones en una sola celda**, lo que nos da la posibilidad de agrupar cálculos, definir elementos y estructurar mejor nuestro código. ¡Incluso podemos escribir programas completos en una sola celda!

Quizás ya te preguntaste al respecto de esto cuando te encontraste con el código de despedida al final del notebook introductorio. ¿Lo recuerdas? Era algo similar a lo siguiente: 

In [11]:
msg = "Espero que te haya gustado la introducción!"
print(msg)

Espero que te haya gustado la introducción!


La anterior es, justamente, una celda de código que contiene múltiples instrucciones. Acá llegamos a una de las características del funcionamiento de Python que deberás tener siempre en mente: **el intérprete de Python considera que cada línea es una instrucción única a menos que se le indique lo contrario**. De este modo, la celda anterior contiene dos instrucciones distintas: una que almacena un mensaje, y otra que lo muestra como resultado de la ejecución.  

Entonces, cuando ejecutamos una celda que contiene varias instrucciones, Python **lee y ejecuta cada línea de forma independiente, de arriba hacia abajo, en orden**, tal como si estuviera siguiendo una lista de pasos. Cada operación se procesa en secuencia antes de pasar a la siguiente, lo que nos permite realizar cálculos intermedios, almacenar resultados y hacer que diferentes partes del código interactúen entre sí.

### Comentarios

Debido a la característica anteriormente descrita, surge la necesidad de poder agregar explicaciones o aclaraciones en texto **entre las instrucciones de una misma celda** para hacer más comprensible el código desarrollado o simplemente mejorar la claridad de lo programado. Para ello, en Python se utiliza el símbolo `#` para indicar que todo el contenido de una instrucción, a partir de ese símbolo, no debe ser ejecutado, dado que corresponde a un **comentario para humanos** que la máquina debe ignorar en su procesamiento. Por ejemplo, podemos utilizar _comentarios_ para clarificar nuestro código anterior: 

In [16]:
# Esta línea es un comentario ignorado por el intérprete 

# La siguiente es la primera instrucción de la celda:
msg = "Espero que te haya gustado la introducción!"

# y esta es la segunda instrucción:
print(msg)

Espero que te haya gustado la introducción!


Estos comentarios incluso se pueden utilizar luego de las instrucciones de una misma línea:

In [1]:
356*24   # la cantidad de horas en un año:

8544

Como recomendación general, incluir **comentarios explicativos** es una muy buena práctica al programar. Es decir, es importante tomarse el tiempo de incluir anotaciones en comentarios que expliquen qué hace cada parte del programa. Los comentarios no afectan la ejecución del código, pero son fundamentales para hacerlo más claro y comprensible. Agregar explicaciones breves sobre cálculos, decisiones o estructuras complejas **te ahorrará mucho tiempo cuando necesites revisar tus desarrollos en el futuro**. También facilita que otras personas puedan entender y reutilizar tú código sin necesidad de tener que analizar cada línea en detalle. Un código bien comentado no solo es más fácil de leer, sino que también ayuda a detectar errores y mantener un flujo de trabajo organizado.

#### Resultados de múltiples instrucciones

Es importante notar que, por defecto, **el resultado que se muestra al ejecutar una celda con múltiples instrucciones suele ser la salida de la última instrucción escrita**. Revisa el siguiente ejemplo:

In [17]:
# ¿El resuldato de cuál de estas operaciones se muestra? 
4+5
6*8

48

Si tienes dudas, puedes probar _comentando_ la última línea de la celda anterior para verificar que Python está igualmente realizando el cálculo, aunque no esté mostrando su resultado en pantalla. 

Por el contrario, si necesitamos mostrar otros valores intermedios dentro de la misma celda (sin tener que _comentar_ las líneas siguientes), debemos usar la _función_ `print()`. Como ya pudiste ver en el notebook anterior, esta instrucción nos permite mostrar explícitamente cualquier información en cualquier punto del código. Esto es especialmente útil cuando queremos revisar los pasos de un cálculo o analizar el comportamiento del programa. 

La función `print()` muestra en pantalla el elemento que incluyas entre sus paréntesis. Ya hemos visto que para mostrar un texto cualquiera con esta función es necesario encerrarlo con comillas. En el caso de resultados de operaciones, podemos usar igualmente `print()` agregando la operación (o su resultado) dentro de los paréntesis. Por ejemplo:

In [24]:
# Estas tres instrucciones se mostrarán:
print("Este es el resultado de 2+2:")
print(2+2)
print("Listo!")

Este es el resultado de 2+2:
4
Listo!


**EJERCICIO:** A continuación, tienes una celda con múltiples instrucciones incluyendo operaciones matemáticas para que la modifiques a tu gusto: comenta la(s) línea(s) de tu elección, agrega comentarios nuevos o agrega/elimina instrucciones `print()` para que practiques lo estudiado hasta ahora. ¡Ejecútala luego de cada cambio para comprobar qué resultado se muestra al final! ¿Puedes conseguir que se muestre el resultado de la operación $2^3$ sin eliminar los mensajes posteriores?

In [21]:
# recuerda que puedes gregar todos los comentarios que quieras!
2**3

# mostremos un mensaje:
print("Este es un mensaje en medio de la ejecución.")

4+5

5*2

Este es un mensaje en medio de la ejecución.


10

## 2.3 Creando Elementos

Hasta ahora, hemos usado Python para realizar cálculos y ejecutar instrucciones individuales. Pero en programación, no siempre queremos trabajar con valores sueltos, sino que necesitamos **crear y gestionar información dentro del programa**. A continuación, daremos un paso más y aprenderemos a definir nuestros propios elementos digitales para luego utilizarlos en distintas operaciones. Iniciaremos con los elementos más sencillos: números y texto, pero lo que aprendas acá se aplicará a todos los elementos complejos que veremos más adelante. A medida que avancemos, veremos cómo Python reconoce y clasifica lo que creamos, y descubriremos detalles clave sobre cómo estos elementos existen y son gestionados dentro del programa. A partir de ahora, utilizaremos el término **objeto** para referirnos a los elementos digitales que existen dentro de un programa. Indagaremos bastante sobre este concepto mientras vayamos avanzando en este tutorial.

### 2.3.1 Objetos numéricos: enteros y flotantes

En Python, podemos **crear objetos numéricos** de manera sencilla utilizando el símbolo `=`. Este símbolo es un _operador_ que nos permite indicarle al intérprete que le **asigne un valor a un nombre** que nos permitirá poder usar ese valor más adelante. Veamos un ejemplo:

In [2]:
# Acá tenemos dos definiciones
a = 2
b = 3.1415

Al ejecutar la celda anterior, el intérprete creará ambos objetos (2 y 3.1415) y registrará sus nombres (a, b) de forma persistente hasta que sean eliminados manualmente o el kernel se reinicie. Estos nombres, los cuales llamaremos **referencias**, se pueden utilizar para acceder y operar con los objetos creados. Por ejemplo, acá podrás ver el resultado de algunas operaciones:

In [8]:
a+b

5.141500000000001

In [9]:
a*b

6.283

In [10]:
a**b

8.824411082479122

La creación de objetos y sus referencias con el `=` también admite el uso de otras referencias a objetos anteriomente. Por ejemplo: 

In [11]:
c = a*b
print(c)

6.283


**EJERCICIO**: Combina lo que hemos estudiado hasta ahora creando una secuencia de instrucciones que: definan algunos objetos numéricos, realicen operaciones matemáticas entre ellos, almacenen resultados en nuevos objetos y muestren varios mensajes de los resultados parciales de tu código. Crea, al menos, 5 líneas de Python (que no sean comentarios) que implementen estas acciones.

In [ ]:
# Completa el ejercicio en esta celda




Entonces, resumiendo lo anterior, cuando programamos: 

> a = 2

no le estamos indicando al intérprete que "guarde" el número 2 dentro de `a`. En realidad, Python crea el número 2 **como un objeto independiente**, y luego le asigna **una referencia** (`a`) que nos permite realizar operaciones con él.

Esto significa que la referencia y el objeto son **dos cosas distintas**: la referencia es simplemente un "etiqueta" que apunta a un valor representado digitalmente en alguna parte de la memoria de la computadora. Este concepto es clave para entender cómo Python maneja los datos y cómo podemos manipularlos de manera eficiente.


#### Tipos de los objetos

Una característica importante de los objetos en Python es que todos tienen asociado un **tipo** que define qué tipo de información representa el elemento digital y qué operaciones podemos realizar con él. Por ejemplo, los números enteros (`int`) y los números decimales o de punto flotante (`float`) son dos tipos distintos, aunque ambos representen valores numéricos. Si bien a primera vista parece _innecesario_ tener dos (o más) tipos para tratar con los números, es importante que consideres que las operaciones internas que realiza la computadora para poder hacer los cálculos son distintas dependiendo del tipo de objeto numérico con que esté trabajando. Si bien esta característica ha sido una limitación importante a lo largo de la historia de la programación, felizmenete Python puede sobrellevarla sin mayor intervención por nuestra parte, de modo que no indagaremos en ello por ahora mientras estemos estudiando lo básico del lenguaje, sino que nos centraremos en identificar todos los tipos de los objetos que utilicemos.

Para averiguar a qué tipo pertenece un objeto particular, se utiliza la función `type()` que nos entregará esta información de manera sencilla. Comprobemos los tipos de los objetos definidos con anterioridad:

In [20]:
# a = 2
type(a)

int

In [18]:
# b= 3.1415
type(b)

float

Vemos que, utilizando sus referencias, es posible consultar los tipos de nuestros objetos creados, los cuales efectivamente corresponden a clases distintas.

Python es conocido por ser un lenguaje _con orientación a objetos_, es decir, utiliza los objetos como herramienta principal para funcionar. Para que veamos esto en acción, podemos preguntar el tipo de un objeto "sin referencia" como el simple número 7:

In [19]:
type(7)

int

Vemos que Python reconoció el tipo del objeto que representa el número $7$ a pesar que no le asignamos ninguna referencia. Es decir, **Python siempre creará los objetos que necesite para poder ejecutar las instrucciones que hayas programado**, incluso aunque estos no se sigan utilizando luego (es decir, no se les asignó referencia).

En adelante, utilizaremos `type` para conocer la clase de los objetos de tipo nuevo que vayamos estudiando. Ten en cuenta que cada vez que sepas realmente qué tipo de objeto está asignado a una referencia, puedes utilizar esta misma función para determinarlo.

### 2.3.2 Cadenas de caracteres (**strings**)

Como ya lo hemos visto en algunos ejemplos anteriores, Python también permite definir objetos que contienen _secuencias de caracteres alfanuméricos_ que nos permiten almacenar textos. Estos objetos son conocidos como cadenas o **strings**, y su tipo es `str`. Las strings permiten almacenar texto de largo arbitrario en estas secuencias de caracteres, y una vez definidos ¡podemos crear instrucciones para realizar operaciones con ellas tal como con los objetos numéricos!

Para definir una string se utiliza una notación especial que le indica al intérprete donde comienza y termina el "texto" que la compone, para así diferenciarlo del resto de las instrucciones programadas: el texto de la string debe encerrarse entre comillas dobles `" "` o simples `' '`. De este modo, de forma análoga a los objetos numéricos, podemos definir una string de la forma:

In [ ]:
s1 = "Hola"

Verifiquemos el tipo de este objeto con la función `type()`:

In [ ]:
# Tipo del objeto que referencia s1:
type(s1)

str

#### Representación de cadenas de texto

En este punto debemos clarificar el uso de la función `print()`, que hasta ahora hemos utilizado instintivamente.

La función `print()` en Python nos permite **mostrar información en un formato legible para humanos**, pero es importante notar que el texto que genera puede verse diferente al resultado que aparece automáticamente cuando ejecutamos una celda. Esto ocurre porque `print()` convierte los valores a una representación de texto más amigable, eliminando detalles técnicos que Python muestra cuando devuelve directamente un objeto.

Por ejemplo, al ejecutar una celda con un string, Python muestra su representación formal, incluyendo las comillas que lo delimitan:

In [21]:
s2 = "Texto de ejemplo"

s2

'Texto de ejemplo'

En cambio, si usamos `print()`, veremos solo el contenido del texto sin las comillas adicionales:

In [22]:
print(s2)

Texto de ejemplo


Esto adquiere mayor importancia cuando consideramos texto que contiene caracteres que representan alguna característica del formato del mismo, por ejemplo, el caracter popular _salto de línea_: `\n`.

In [24]:
# Definimos una string con el caracter especial de salto de linea:
s3 = "una linea \n otra linea"

In [25]:
# Su representación automática:
s3

'una linea \n otra linea'

In [27]:
# Su representación en texto real (para humanos):
print(s3)

una linea 
 otra linea


Esta diferencia se debe a que, cuando una celda ejecuta una última instrucción sin `print()`, Python muestra la **representación interna** del objeto, útil para analizar y corregir un código en desarrollo. En cambio, `print()` genera una salida más limpia y fácil de leer, pensada para interactuar con el usuario.

**CONCLUSIÓN**  
Puedes utilizar la representación automática de las celdas de texto para mostrar valores de objetos mientras estés desarrollando tus programas. ¡Es una forma fácil y rápida de verificar que tus instrucciones están correctas!. En cambio, cuando necesites mostrar valores **como parte del funcionamiento final** de tu programa o conjunto de instrucciones, **siempre prefiere utilizar `print()`**. Esta última función es el estándar para mostrar información en un programa en Python. 

#### Operaciones con strings

Las strings se pueden combinar ya sea en una misma instrucción o utilizando sus referencias:

In [ ]:
"Hola" "mundo"

'Holamundo'

In [ ]:
# Las strings se pueden combinar (por sus referencias) con el operador +
s3 = "Hola"
s4 = " mundo"
s3 + s4

'Hola mundo'

En general, al aparece un operador (como el +) que requiere realizar alguna operación entre distintos objetos, Python le pregunta a la implementación de los tipos (clases) de esos objetos para saber si puede utilizar este operador entre ellos.

En el caso de las strings, estas se puden sumar entre si, lo cual se traduce en una **concatenación**, que no es equivalente a la suma algebraica.

Podemos probar otros operadores:

In [ ]:
# La multiplicación me concatena múltiples veces una string.
3*"Hola"

'HolaHolaHola'

In [ ]:
# Se posible combinar estos operadores:
s5 = "Ven"
s6 = " michi"
print(s5+3*s6+"!")

Ven michi michi michi!


In [ ]:
# El operador *, al igual que en su versión algebraica,
# respecta el orden y jerarquía
print(s5+s6*3+"!")

Ven michi michi michi!


Las strings, en el fondo, se pueden entender como arrays de números enteros, donde cada caracter tiene un equivalente numérico definido por la tabla ASCII.

[Ejemplo de tabla ASCII de San Google](https://elcodigoascii.com.ar/)

Al ser arrays, puedo acceder a los elementos individuales del lenguaje. Para ello, utilizamos el **operador de indexación [ ]**.

In [ ]:
# Ejemplo, para acceder a los elementos del "Hola",
# cuya referencia es s1:
s1[0], s1[1], s1[2], s1[3]

('H', 'o', 'l', 'a')

Este operador me retorna los elementos de mi objetos según su índice, **partiendo desde cero**.
* En Python, todas las cosas indexadas parten desde cero
* Y el operador de indexación aplica la misma lógica para todos los objetos (fundamentales) del lenguaje.

In [ ]:
# Tambien admite números negativos:
s1[-1]

'a'

Con los números negativos accedemos a los elementos pero contando desde el último (-1) hacia atrás (-2, -3, ...)

In [ ]:
s2[-2], s2[-3]

('e', 'n')

¿qué pasa si le solicito un índice que no corresponde (porque es mayor o menor que los números permitidos según el tamaño del string)?

In [ ]:
s1[4]

IndexError: string index out of range

El operador de indexación permite realizar una operación que combina elementos consecutivos del objeto, creando uno nuevo: esto es la **operación slice**. Para ello, utilizamos el ":" dentro del operador y dos enteros:

In [ ]:
s7 = s5+3*s6+"!"
print(s7)

Ven michi michi michi!


In [ ]:
s7[4:9]

'michi'

Vemos que el primer número de la indexación es incluyente, mientras que el segundo es excluyente. Esto funciona de forma análoga en caso de utilizar números negativos:

In [ ]:
s7[-6:-1]

'michi'

Si omitimos uno de los números, se entiende como "hasta el principio" (en el caso del primero) o "hasta el final" (en el caso del último).

In [ ]:
# desde el principio:
s1[:3]

'Hol'

In [ ]:
# hasta el final:
s1[2:]

'la'

La decisión de que uno sea excluyente y otro no, es para poder tener simetría en la operación conjunta:

In [ ]:
ind = 1
s1[:ind] + s1[ind:]

'Hola'

Python permite crear string con múltiples líneas (de las celdas de instrucción) utilizando la comilla triple:

In [ ]:
s7 = """Hola
mundo"""

In [ ]:
print(s7)

Hola
mundo


Las comillas triples tienen un significado en Python porque el intérprete automáticamente las asocia con documentación del código.

### Importante

 A nivel computacional, el problema que conlleva trabajar con texto es que los objetos que lo almacenen necesitan conocer, antes de crearlo, el tamaño final del texto a almacenar para utilizar una cantidad de memoria suficiente para contener toda la información. Por ejemplo: no se necesita la misma cantidad de bytes para almacenar la palabra

> "hola"

en comparación con el texto

> "Hay una tetera entre la Tierra y Marte que gira alrededor del Sol en una órbita elíptica y es tan pequeña que no puede ser vista ni por los telescopios humanos más potentes."

La característica anterior impone que las strings sean objetos **inmutables**, es decir, que _no se pueden modificar una vez que han sido creados_.

In [ ]:
# Intentemos pasar a minúscula el primer elemento de "Hola"
s1[0] = "h"

TypeError: 'str' object does not support item assignment

Si necesitamos modificar una string, el procedimiento es _crear objetos nuevos a partir de los ya definidos, utilizando los operadores vistos hasta el momento_ (=, +, *, ., []).

In [ ]:
s1cap = "H" + s1[1:]
print(s1cap)

Hola


In [21]:
s1 = "Hola"
s2 = "Mundo"

In [22]:
# Puedo combinar las variables utilizando las mismas comillas
"Hola"    " "     "Mundo"

'Hola Mundo'

In [23]:
# Incluso puedo realizar operaciones entre las strings:
s1 + s2

'HolaMundo'

In [24]:
s1 + " " + s2 + " cruel"

'Hola Mundo cruel'

In [25]:
#y eso incluso lo podemos guardar en una variable nueva:
s3 = s1 + " " + s2 + " cruel"

In [26]:
s3

'Hola Mundo cruel'

In [13]:
# Tip: podemos abreviar algunas operaciones cuando hay variables repetidas
a = 5
a = a + 1
a

6

In [14]:
# Esto se puede escribir como
a = 5
a += 1
a

6

In [15]:
# Python "sabe" (porque el intérprete está en ejecución y no se ha detenido)
# qué variables (referencias) se han creado hasta el momento.
# Por ejemplo, si consultamos por una que no hemos definido:
n

NameError: name 'n' is not defined

In [16]:
# Podemos acceder a la lista de referencias
# definidas con el comando mágico:
%who

a	 b	 


In [17]:
# Como sucedión con el ejemplo de la referencia desconocida,
# Python puede reconocer el tipo de error cuando este sucede. Ejemplo:
zero = 0
a / zero

ZeroDivisionError: division by zero